plugins/steelseries/fu-steelseries-fizz.c

```diff
@@ -0,0 +1,720 @@
+					     id,
+					     buf,
+					     bufsz,
+					     fu_progress_get_child(progress),
+					     error)) {
+		g_prefix_error(error, "failed to write file 0x%02x:0x%02x: ", fs, id);
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	if (!fu_steelseries_fizz_file_crc32(device, fs, id, &calculated_crc, &stored_crc, error)) {
+		g_prefix_error(error,
+			       "failed to get file CRC32 from FS 0x%02x ID 0x%02x: ",
+			       fs,
+			       id);
+		return FALSE;
+	}
+	if (calculated_crc != stored_crc)
+		g_warning("%s: checksum mismatch, got 0x%08x, expected 0x%08x",
+			  fu_device_get_name(device),
+			  calculated_crc,
+			  stored_crc);
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_fizz_write_firmware(FuDevice *device,
+				   FuFirmware *firmware,
+				   FuProgress *progress,
+				   FwupdInstallFlags flags,
+				   GError **error)
+{
+	guint16 pid;
+	guint8 fs = 0x01U;
+	guint8 id = 0x05U;
+
+	/* it is a mouse */
+	pid = fu_usb_device_get_pid(FU_USB_DEVICE(device));
+	if (pid == 0x183a) {
+		fs = 0x02U;
+		id = 0x08U;
+	}
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100);
+
+	if (!fu_steelseries_fizz_write_file(device,
+					    fs,
+					    id,
+					    firmware,
+					    fu_progress_get_child(progress),
+					    flags,
+					    error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_fizz_parse_firmware(FuFirmware *firmware, FwupdInstallFlags flags, GError **error)
+{
+	guint32 checksum_tmp;
+	guint32 checksum;
+	g_autoptr(GBytes) blob = NULL;
+
+	blob = fu_firmware_get_bytes(firmware, error);
+	if (blob == NULL)
+		return FALSE;
+
+	if (!fu_common_read_uint32_safe(g_bytes_get_data(blob, NULL),
+					g_bytes_get_size(blob),
+					g_bytes_get_size(blob) - sizeof(checksum),
+					&checksum,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+	checksum_tmp = fu_common_crc32(g_bytes_get_data(blob, NULL),
+				       g_bytes_get_size(blob) - sizeof(checksum_tmp));
+	if (checksum_tmp != checksum) {
+		if ((flags & FWUPD_INSTALL_FLAG_IGNORE_CHECKSUM) == 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "checksum mismatch, got 0x%08x, expected 0x%08x",
+				    checksum_tmp,
+				    checksum);
+			return FALSE;
+		}
+		g_debug("ignoring checksum mismatch, got 0x%08x, expected 0x%08x",
+			checksum_tmp,
+			checksum);
+	}
+
+	fu_firmware_add_flag(firmware, FU_FIRMWARE_FLAG_HAS_CHECKSUM);
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_steelseries_fizz_prepare_firmware(FuDevice *device,
+				     GBytes *fw,
+				     FwupdInstallFlags flags,
+				     GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_firmware_new();
+
+	if (!fu_firmware_parse(firmware, fw, flags, error))
+		return NULL;
+
+	if (!fu_steelseries_fizz_parse_firmware(firmware, flags, error))
+		return NULL;
```

Comment:
I'd rather a FuFizzSteelseriesFirmware object..

Reply:
That is in my pipe. I was lazy to create a specific file for that draft :p